File: clients/python/vaultdotenv/crypto.py
"""Vault-env cryptographic primitives — matches the Node.js implementation exactly."""
from __future__ import annotations

import hashlib
import hmac
import os
import time
from base64 import b64decode, b64encode
from typing import Optional, Union

from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes

SALT_ENCRYPT = b"vault-encrypt-v1"
SALT_AUTH = b"vault-auth-v1"
IV_LENGTH = 12
TAG_LENGTH = 16


def derive_key(vault_key: str, salt: str | bytes, device_secret: str | None = None) -> bytes:
    """Derive a 256-bit key using HKDF. Matches Node.js deriveKey() exactly."""
    if isinstance(salt, str):
        salt = salt.encode()

    if device_secret:
        ikm = hmac.new(vault_key.encode(), device_secret.encode(), hashlib.sha256).digest()
    else:
        ikm = vault_key.encode()

    return HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        info=b"",
    ).derive(ikm)


def encrypt(plaintext: str, vault_key: str, device_secret: str | None = None) -> str:
    """AES-256-GCM encrypt. Returns base64(iv + tag + ciphertext)."""
    key = derive_key(vault_key, SALT_ENCRYPT, device_secret)
    iv = os.urandom(IV_LENGTH)
    aesgcm = AESGCM(key)
    # AESGCM.encrypt returns ciphertext + tag (tag is last 16 bytes)
    ct_with_tag = aesgcm.encrypt(iv, plaintext.encode(), None)
    # Split: ciphertext is everything except last 16 bytes, tag is last 16
    ciphertext = ct_with_tag[:-TAG_LENGTH]
    tag = ct_with_tag[-TAG_LENGTH:]
    # Pack as: iv + tag + ciphertext (matches Node.js format)
    packed = iv + tag + ciphertext
    return b64encode(packed).decode()


def decrypt(encrypted_b64: str, vault_key: str, device_secret: str | None = None) -> str:
    """AES-256-GCM decrypt from base64(iv + tag + ciphertext)."""
    key = derive_key(vault_key, SALT_ENCRYPT, device_secret)
    packed = b64decode(encrypted_b64)
    iv = packed[:IV_LENGTH]
    tag = packed[IV_LENGTH : IV_LENGTH + TAG_LENGTH]
    ciphertext = packed[IV_LENGTH + TAG_LENGTH :]
    aesgcm = AESGCM(key)
    # AESGCM.decrypt expects ciphertext + tag concatenated
    plaintext = aesgcm.decrypt(iv, ciphertext + tag, None)
    return plaintext.decode()


def sign(vault_key: str, body: str, device_secret: str | None = None) -> str:
    """HMAC-SHA256 signature for API auth. Returns 'v=<timestamp>,d=<hex digest>'."""
    auth_key = derive_key(vault_key, SALT_AUTH, device_secret)
    timestamp = str(int(time.time() * 1000))
    message = (body + timestamp).encode()
    digest = hmac.new(auth_key, message, hashlib.sha256).hexdigest()
    return f"v={timestamp},d={digest}"


def parse_vault_key(vault_key: str) -> dict | None:
    """Parse vk_<projectId>_<secret> into components."""
    if not vault_key or not vault_key.startswith("vk_"):
        return None
    parts = vault_key.split("_", 2)
    if len(parts) < 3:
        return None
    return {"project_id": parts[1], "secret": parts[2]}


def generate_device_secret() -> str:
    """Generate a random 32-byte hex device secret."""
    return os.urandom(32).hex()


def hash_device_secret(device_secret: str) -> str:
    """SHA-256 hash of device secret for server storage."""
    return hashlib.sha256(device_secret.encode()).hexdigest()

File: src/crypto.js
'use strict';

const crypto = require('crypto');

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 12;
const TAG_LENGTH = 16;
const SALT_ENCRYPT = 'vault-encrypt-v1';
const SALT_AUTH = 'vault-auth-v1';
/**
 * Derive a 256-bit key from the vault key using HKDF.
 * When deviceSecret is provided, the key material is HMAC(vaultKey, deviceSecret)
 * so both pieces are required to derive the same encryption/auth keys.
 *
 * @param {string} vaultKey - The master vault key
 * @param {string} salt - Purpose-specific salt
 * @param {string} [deviceSecret] - Device-bound secret (from ~/.vault/device.key)
 * @returns {Buffer} 32-byte derived key
 */
function deriveKey(vaultKey, salt, deviceSecret) {
  const ikm = deviceSecret
    ? crypto.createHmac('sha256', vaultKey).update(deviceSecret).digest()
    : vaultKey;
  return crypto.hkdfSync('sha256', ikm, salt, '', 32);
}

/**
 * Generate a random device secret.
 * @returns {string} 64-char hex string
 */
function generateDeviceSecret() {
  return crypto.randomBytes(32).toString('hex');
}

/**
 * Hash a device secret for server-side storage (server never sees the raw secret).
 * @param {string} deviceSecret - The raw device secret
 * @returns {string} hex-encoded SHA-256 hash
 */
function hashDeviceSecret(deviceSecret) {
  return crypto.createHash('sha256').update(deviceSecret).digest('hex');
}

/**
 * Encrypt a plaintext string with AES-256-GCM.
 * @param {string} plaintext - Data to encrypt
 * @param {string} vaultKey - Master vault key
 * @param {string} [deviceSecret] - Device-bound secret
 * @returns {string} Base64-encoded iv:tag:ciphertext
 */
function encrypt(plaintext, vaultKey, deviceSecret) {
  const key = deriveKey(vaultKey, SALT_ENCRYPT, deviceSecret);
  const iv = crypto.randomBytes(IV_LENGTH);
  const cipher = crypto.createCipheriv(ALGORITHM, key, iv);

  let encrypted = cipher.update(plaintext, 'utf8');
  encrypted = Buffer.concat([encrypted, cipher.final()]);
  const tag = cipher.getAuthTag();

  // Pack as: iv + tag + ciphertext (all base64)
  const packed = Buffer.concat([iv, tag, encrypted]);
  return packed.toString('base64');
}

/**
 * Decrypt an AES-256-GCM encrypted string.
 * @param {string} encryptedBase64 - Base64-encoded iv:tag:ciphertext
 * @param {string} vaultKey - Master vault key
 * @param {string} [deviceSecret] - Device-bound secret
 * @returns {string} Decrypted plaintext
 */
function decrypt(encryptedBase64, vaultKey, deviceSecret) {
  const key = deriveKey(vaultKey, SALT_ENCRYPT, deviceSecret);
  const packed = Buffer.from(encryptedBase64, 'base64');

  const iv = packed.subarray(0, IV_LENGTH);
  const tag = packed.subarray(IV_LENGTH, IV_LENGTH + TAG_LENGTH);
  const ciphertext = packed.subarray(IV_LENGTH + TAG_LENGTH);

  const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
  decipher.setAuthTag(tag);

  let decrypted = decipher.update(ciphertext);
  decrypted = Buffer.concat([decrypted, decipher.final()]);
  return decrypted.toString('utf8');
}

/**
 * Generate HMAC signature for API authentication.
 * @param {string} vaultKey - Master vault key
 * @param {string} body - Request body string
 * @param {string} [deviceSecret] - Device-bound secret
 * @returns {{ signature: string, timestamp: string }}
 */
function sign(vaultKey, body, deviceSecret) {
  const authKey = deriveKey(vaultKey, SALT_AUTH, deviceSecret);
  const timestamp = String(Date.now());
  const input = body + timestamp;
  const digest = crypto.createHmac('sha256', authKey).update(input, 'utf8').digest('hex');
  return { signature: `v=${timestamp},d=${digest}`, timestamp };
}

/**
 * Verify an HMAC signature.
 * @param {string} vaultKey - Master vault key
 * @param {string} body - Request body string
 * @param {string} sigHeader - Signature header value
 * @param {number} [maxAgeMs=300000] - Maximum age in ms (default 5 min)
 * @param {string} [deviceSecret] - Device-bound secret
 * @returns {{ valid: boolean, reason?: string }}
 */
function verify(vaultKey, body, sigHeader, maxAgeMs = 300_000, deviceSecret) {
  if (!vaultKey || !body || !sigHeader) return { valid: false, reason: 'missing_params' };

  const parts = {};
  for (const part of sigHeader.split(',')) {
    const idx = part.indexOf('=');
    if (idx !== -1) parts[part.slice(0, idx).trim()] = part.slice(idx + 1).trim();
  }

  const timestamp = parts['v'];
  const providedDigest = parts['d'];
  if (!timestamp || !providedDigest) return { valid: false, reason: 'malformed_signature' };

  const age = Date.now() - parseInt(timestamp, 10);
  if (isNaN(age) || age < -60_000 || age > maxAgeMs) return { valid: false, reason: 'timestamp_stale' };

  const authKey = deriveKey(vaultKey, SALT_AUTH, deviceSecret);
  const input = body + timestamp;
  const computed = crypto.createHmac('sha256', authKey).update(input, 'utf8').digest('hex');

  const valid = crypto.timingSafeEqual(
    Buffer.from(computed, 'hex'),
    Buffer.from(providedDigest, 'hex'),
  );

  return { valid };
}

/**
 * Parse a vault key into its components.
 * Format: vk_<projectId>_<secret>
 * @param {string} vaultKey
 * @returns {{ projectId: string, secret: string } | null}
 */
function parseVaultKey(vaultKey) {
  if (!vaultKey || !vaultKey.startsWith('vk_')) return null;
  const parts = vaultKey.split('_');
  if (parts.length < 3) return null;
  return {
    projectId: parts[1],
    secret: parts.slice(2).join('_'),
  };
}

/**
 * Generate a new vault key for a project.
 * @param {string} projectId
 * @returns {string} vk_<projectId>_<random>
 */
function generateVaultKey(projectId) {
  const secret = crypto.randomBytes(32).toString('hex');
  return `vk_${projectId}_${secret}`;
}

module.exports = {
  encrypt,
  decrypt,
  sign,
  verify,
  parseVaultKey,
  generateVaultKey,
  generateDeviceSecret,
  hashDeviceSecret,
  deriveKey,
};
